File: server/routes/leaderboards.js
/**
 * Leaderboard Routes
 * ===================
 * Sharded leaderboard system designed for 1M+ players.
 * Uses Redis sorted sets with shard merging for queries.
 * Extended with season-aware, regional, and multiplayer rankings.
 *
 * GET /leaderboards/:gameId          - Global leaderboard for a game
 * GET /leaderboards/:gameId/me       - Current player's rank
 * GET /leaderboards/:gameId/around   - Nearby ranks around player
 * GET /leaderboards/:gameId/friends  - Friend leaderboard
 * GET /leaderboards/:gameId/ranked   - Season/region multiplayer leaderboard
 * GET /leaderboards/global           - Aggregate across all games
 * GET /leaderboards/seasons          - List all seasons
 * GET /leaderboards/seasons/current  - Current active season
 * POST /leaderboards/submit-match    - Submit multiplayer match result
 */

const express = require('express');
const db = require('../models/db');
const cache = require('../services/cache');
const config = require('../config');

const router = express.Router();

/**
 * Get leaderboard for a specific game.
 * Tries Redis cache first, falls back to PostgreSQL.
 */
router.get('/:gameId', async (req, res, next) => {
    try {
        const { gameId } = req.params;
        const tenantId = req.tenantId;
        const limit = Math.min(parseInt(req.query.limit, 10) || 50, 200);
        const offset = parseInt(req.query.offset, 10) || 0;
        const period = req.query.period || 'all';

        // Try cache first
        const cacheKey = `lb_result:${tenantId}:${gameId}:${period}:${offset}:${limit}`;
        const cached = await cache.get(cacheKey);
        if (cached) {
            return res.json(cached);
        }

        // Fallback to DB
        let query;
        const params = [tenantId, gameId, limit, offset];

        if (period === 'all') {
            query = `
                SELECT gp.player_id, gp.high_score, gp.stars, gp.level, gp.play_count,
                       p.display_name, p.avatar_character,
                       RANK() OVER (ORDER BY gp.high_score DESC) as rank
                FROM game_progress gp
                JOIN players p ON p.id = gp.player_id AND p.tenant_id = gp.tenant_id
                WHERE gp.tenant_id = $1 AND gp.game_id = $2 AND gp.high_score > 0
                ORDER BY gp.high_score DESC
                LIMIT $3 OFFSET $4
            `;
        } else {
            // Time-filtered: use score_history
            const periodFilter = _getPeriodFilter(period);
            query = `
                SELECT sh.player_id, MAX(sh.score) as high_score,
                       p.display_name, p.avatar_character,
                       RANK() OVER (ORDER BY MAX(sh.score) DESC) as rank
                FROM score_history sh
                JOIN players p ON p.id = sh.player_id AND p.tenant_id = sh.tenant_id
                WHERE sh.tenant_id = $1 AND sh.game_id = $2 AND sh.created_at >= $5
                GROUP BY sh.player_id, p.display_name, p.avatar_character
                ORDER BY high_score DESC
                LIMIT $3 OFFSET $4
            `;
            params.push(periodFilter);
        }

        const result = await db.query(query, params);

        // Count total entries
        const countResult = await db.query(
            'SELECT COUNT(*) FROM game_progress WHERE tenant_id = $1 AND game_id = $2 AND high_score > 0',
            [tenantId, gameId]
        );

        const response = {
            gameId,
            period,
            total: parseInt(countResult.rows[0].count, 10),
            offset,
            limit,
            entries: result.rows.map(row => ({
                rank: parseInt(row.rank, 10),
                playerId: row.player_id,
                displayName: row.display_name,
                avatarCharacter: row.avatar_character,
                highScore: parseInt(row.high_score, 10),
                stars: row.stars,
                level: row.level,
                playCount: row.play_count
            }))
        };

        // Cache for configured duration
        await cache.set(cacheKey, response, config.leaderboard.cacheSeconds);

        res.json(response);
    } catch (err) {
        next(err);
    }
});

/**
 * Get current player's rank.
 */
router.get('/:gameId/me', async (req, res, next) => {
    try {
        if (!req.player) {
            return res.status(401).json({ error: 'Authentication required' });
        }

        const { gameId } = req.params;
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;

        const result = await db.query(`
            SELECT player_id, high_score, stars, level, play_count,
                   (SELECT COUNT(*) + 1 FROM game_progress
                    WHERE tenant_id = $1 AND game_id = $2 AND high_score > gp.high_score) as rank
            FROM game_progress gp
            WHERE player_id = $3 AND tenant_id = $1 AND game_id = $2
        `, [tenantId, gameId, playerId]);

        if (result.rows.length === 0) {
            return res.json({ gameId, rank: null, highScore: 0 });
        }

        const row = result.rows[0];
        res.json({
            gameId,
            rank: parseInt(row.rank, 10),
            highScore: parseInt(row.high_score, 10),
            stars: row.stars,
            level: row.level,
            playCount: row.play_count
        });
    } catch (err) {
        next(err);
    }
});

/**
 * Get leaderboard entries around the current player.
 */
router.get('/:gameId/around', async (req, res, next) => {
    try {
        if (!req.player) {
            return res.status(401).json({ error: 'Authentication required' });
        }

        const { gameId } = req.params;
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;
        const range = Math.min(parseInt(req.query.range, 10) || 5, 25);

        // Get player's score first
        const playerResult = await db.query(
            'SELECT high_score FROM game_progress WHERE player_id = $1 AND tenant_id = $2 AND game_id = $3',
            [playerId, tenantId, gameId]
        );

        if (playerResult.rows.length === 0) {
            return res.json({ gameId, entries: [] });
        }

        const playerScore = parseInt(playerResult.rows[0].high_score, 10);

        // Get players around this score
        const result = await db.query(`
            (SELECT gp.player_id, gp.high_score, gp.stars, gp.level,
                    p.display_name, p.avatar_character,
                    RANK() OVER (ORDER BY gp.high_score DESC) as rank
             FROM game_progress gp
             JOIN players p ON p.id = gp.player_id AND p.tenant_id = gp.tenant_id
             WHERE gp.tenant_id = $1 AND gp.game_id = $2 AND gp.high_score >= $3
             ORDER BY gp.high_score ASC
             LIMIT $4)
            UNION ALL
            (SELECT gp.player_id, gp.high_score, gp.stars, gp.level,
                    p.display_name, p.avatar_character,
                    RANK() OVER (ORDER BY gp.high_score DESC) as rank
             FROM game_progress gp
             JOIN players p ON p.id = gp.player_id AND p.tenant_id = gp.tenant_id
             WHERE gp.tenant_id = $1 AND gp.game_id = $2 AND gp.high_score < $3
             ORDER BY gp.high_score DESC
             LIMIT $4)
            ORDER BY high_score DESC
        `, [tenantId, gameId, playerScore, range]);

        res.json({
            gameId,
            entries: result.rows.map(row => ({
                playerId: row.player_id,
                displayName: row.display_name,
                avatarCharacter: row.avatar_character,
                highScore: parseInt(row.high_score, 10),
                stars: row.stars,
                level: row.level,
                isCurrentPlayer: row.player_id === playerId
            }))
        });
    } catch (err) {
        next(err);
    }
});

/**
 * Global leaderboard (aggregate total score across all games).
 */
router.get('/global', async (req, res, next) => {
    try {
        const tenantId = req.tenantId;
        const limit = Math.min(parseInt(req.query.limit, 10) || 50, 200);
        const offset = parseInt(req.query.offset, 10) || 0;

        const cacheKey = `lb_result:${tenantId}:global:${offset}:${limit}`;
        const cached = await cache.get(cacheKey);
        if (cached) return res.json(cached);

        const result = await db.query(`
            SELECT p.id as player_id, p.display_name, p.avatar_character,
                   p.total_score, p.games_played,
                   RANK() OVER (ORDER BY p.total_score DESC) as rank
            FROM players p
            WHERE p.tenant_id = $1 AND p.total_score > 0
            ORDER BY p.total_score DESC
            LIMIT $2 OFFSET $3
        `, [tenantId, limit, offset]);

        const countResult = await db.query(
            'SELECT COUNT(*) FROM players WHERE tenant_id = $1 AND total_score > 0',
            [tenantId]
        );

        const response = {
            total: parseInt(countResult.rows[0].count, 10),
            offset,
            limit,
            entries: result.rows.map(row => ({
                rank: parseInt(row.rank, 10),
                playerId: row.player_id,
                displayName: row.display_name,
                avatarCharacter: row.avatar_character,
                totalScore: parseInt(row.total_score, 10),
                gamesPlayed: row.games_played
            }))
        };

        await cache.set(cacheKey, response, config.leaderboard.cacheSeconds);
        res.json(response);
    } catch (err) {
        next(err);
    }
});

// =========================================
// Friend Leaderboard
// =========================================

router.get('/:gameId/friends', async (req, res, next) => {
    try {
        if (!req.player) {
            return res.status(401).json({ error: 'Authentication required' });
        }

        const { gameId } = req.params;
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;

        const result = await db.query(`
            SELECT gp.player_id, gp.high_score, gp.stars, gp.level, gp.play_count,
                   p.display_name, p.avatar_character,
                   RANK() OVER (ORDER BY gp.high_score DESC) as rank
            FROM game_progress gp
            JOIN players p ON p.id = gp.player_id AND p.tenant_id = gp.tenant_id
            WHERE gp.tenant_id = $1 AND gp.game_id = $2 AND gp.high_score > 0
                AND (gp.player_id = $3 OR gp.player_id IN (
                    SELECT CASE WHEN f.player_id = $3 THEN f.friend_id ELSE f.player_id END
                    FROM friendships f
                    WHERE f.tenant_id = $1 AND f.status = 'accepted'
                        AND (f.player_id = $3 OR f.friend_id = $3)
                ))
            ORDER BY gp.high_score DESC
            LIMIT 50
        `, [tenantId, gameId, playerId]);

        res.json({
            gameId,
            entries: result.rows.map(row => ({
                rank: parseInt(row.rank, 10),
                playerId: row.player_id,
                displayName: row.display_name,
                avatarCharacter: row.avatar_character,
                highScore: parseInt(row.high_score, 10),
                stars: row.stars,
                isCurrentPlayer: row.player_id === playerId
            }))
        });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Ranked / Season Multiplayer Leaderboard
// =========================================

router.get('/:gameId/ranked', async (req, res, next) => {
    try {
        const { gameId } = req.params;
        const tenantId = req.player?.tenantId || req.tenantId || 'stem_default';
        const region = req.query.region || 'global';
        const limit = Math.min(parseInt(req.query.limit, 10) || 50, 100);
        const offset = parseInt(req.query.offset, 10) || 0;

        // Get current season
        const activeSeason = await db.query(
            'SELECT id, name FROM seasons WHERE tenant_id = $1 AND is_active = TRUE ORDER BY starts_at DESC LIMIT 1',
            [tenantId]
        );
        const seasonId = activeSeason.rows[0]?.id || null;

        const cacheKey = `lb_ranked:${tenantId}:${gameId}:${region}:${seasonId}:${offset}:${limit}`;
        const cached = await cache.get(cacheKey);
        if (cached) return res.json(cached);

        let query, params;
        if (seasonId) {
            query = `
                SELECT le.player_id, le.score, le.wins, le.losses, le.draws,
                       le.matches_played, le.skill_rating,
                       p.display_name, p.avatar_character,
                       ROW_NUMBER() OVER (ORDER BY le.skill_rating DESC, le.score DESC) as rank
                FROM leaderboard_entries le
                JOIN players p ON p.id = le.player_id AND p.tenant_id = le.tenant_id
                WHERE le.tenant_id = $1 AND le.game_id = $2
                    AND le.region = $3 AND le.season_id = $4
                ORDER BY le.skill_rating DESC, le.score DESC
                LIMIT $5 OFFSET $6
            `;
            params = [tenantId, gameId, region, seasonId, limit, offset];
        } else {
            query = `
                SELECT le.player_id, le.score, le.wins, le.losses, le.draws,
                       le.matches_played, le.skill_rating,
                       p.display_name, p.avatar_character,
                       ROW_NUMBER() OVER (ORDER BY le.skill_rating DESC, le.score DESC) as rank
                FROM leaderboard_entries le
                JOIN players p ON p.id = le.player_id AND p.tenant_id = le.tenant_id
                WHERE le.tenant_id = $1 AND le.game_id = $2
                    AND le.region = $3 AND le.season_id IS NULL
                ORDER BY le.skill_rating DESC, le.score DESC
                LIMIT $4 OFFSET $5
            `;
            params = [tenantId, gameId, region, limit, offset];
        }

        const result = await db.query(query, params);

        const response = {
            gameId,
            region,
            seasonId,
            seasonName: activeSeason.rows[0]?.name || null,
            offset,
            limit,
            entries: result.rows.map(row => ({
                rank: parseInt(row.rank, 10),
                playerId: row.player_id,
                displayName: row.display_name,
                avatarCharacter: row.avatar_character,
                score: parseInt(row.score, 10),
                skillRating: row.skill_rating,
                wins: row.wins,
                losses: row.losses,
                draws: row.draws,
                matchesPlayed: row.matches_played
            }))
        };

        await cache.set(cacheKey, response, 15); // shorter cache for ranked
        res.json(response);
    } catch (err) {
        next(err);
    }
});

// =========================================
// Seasons
// =========================================

router.get('/seasons', async (req, res, next) => {
    try {
        const tenantId = req.player?.tenantId || req.tenantId || 'stem_default';
        const result = await db.query(
            'SELECT * FROM seasons WHERE tenant_id = $1 ORDER BY starts_at DESC LIMIT 20',
            [tenantId]
        );
        res.json({ seasons: result.rows });
    } catch (err) {
        next(err);
    }
});

router.get('/seasons/current', async (req, res, next) => {
    try {
        const tenantId = req.player?.tenantId || req.tenantId || 'stem_default';
        const result = await db.query(
            'SELECT * FROM seasons WHERE tenant_id = $1 AND is_active = TRUE ORDER BY starts_at DESC LIMIT 1',
            [tenantId]
        );
        res.json({ season: result.rows[0] || null });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Submit Multiplayer Match Result
// =========================================

router.post('/submit-match', async (req, res, next) => {
    try {
        if (!req.player) {
            return res.status(401).json({ error: 'Authentication required' });
        }

        const tenantId = req.player.tenantId;
        const { gameId, matchId, results, ratingChanges } = req.body;

        if (!gameId || !results || !Array.isArray(results)) {
            return res.status(400).json({ error: 'gameId and results array required' });
        }

        // Get current season
        const activeSeason = await db.query(
            'SELECT id FROM seasons WHERE tenant_id = $1 AND is_active = TRUE LIMIT 1',
            [tenantId]
        );
        const seasonId = activeSeason.rows[0]?.id || null;

        for (const result of results) {
            const isWin = result.isWinner;
            const isLoss = !isWin && results.length > 1;
            const isDraw = !isWin && results.filter(r => r.placement === result.placement).length > 1;
            const newRating = ratingChanges?.[result.id]?.newRating || null;

            // Upsert global leaderboard entry
            await db.query(`
                INSERT INTO leaderboard_entries
                    (tenant_id, player_id, game_id, season_id, region, score, wins, losses, draws, matches_played, skill_rating)
                VALUES ($1, $2, $3, $4, 'global', $5, $6, $7, $8, 1, COALESCE($9, 1000))
                ON CONFLICT (tenant_id, player_id, game_id, season_id, region)
                DO UPDATE SET
                    score = leaderboard_entries.score + EXCLUDED.score,
                    wins = leaderboard_entries.wins + EXCLUDED.wins,
                    losses = leaderboard_entries.losses + EXCLUDED.losses,
                    draws = leaderboard_entries.draws + EXCLUDED.draws,
                    matches_played = leaderboard_entries.matches_played + 1,
                    skill_rating = COALESCE($9, leaderboard_entries.skill_rating),
                    updated_at = NOW()
            `, [tenantId, result.id, gameId, seasonId,
                result.score || 0, isWin ? 1 : 0, isLoss ? 1 : 0, isDraw ? 1 : 0,
                newRating]);

            // Upsert regional entry
            const playerRegion = await db.query(
                'SELECT region FROM players WHERE id = $1 AND tenant_id = $2',
                [result.id, tenantId]
            );
            const region = playerRegion.rows[0]?.region || 'us-east';

            await db.query(`
                INSERT INTO leaderboard_entries
                    (tenant_id, player_id, game_id, season_id, region, score, wins, losses, draws, matches_played, skill_rating)
                VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, 1, COALESCE($10, 1000))
                ON CONFLICT (tenant_id, player_id, game_id, season_id, region)
                DO UPDATE SET
                    score = leaderboard_entries.score + EXCLUDED.score,
                    wins = leaderboard_entries.wins + EXCLUDED.wins,
                    losses = leaderboard_entries.losses + EXCLUDED.losses,
                    draws = leaderboard_entries.draws + EXCLUDED.draws,
                    matches_played = leaderboard_entries.matches_played + 1,
                    skill_rating = COALESCE($10, leaderboard_entries.skill_rating),
                    updated_at = NOW()
            `, [tenantId, result.id, gameId, seasonId, region,
                result.score || 0, isWin ? 1 : 0, isLoss ? 1 : 0, isDraw ? 1 : 0,
                newRating]);

            // Update player aggregate stats
            await db.query(`
                UPDATE players SET
                    mp_wins = mp_wins + $1, mp_losses = mp_losses + $2,
                    mp_draws = mp_draws + $3, mp_matches = mp_matches + 1,
                    skill_rating = COALESCE($4, skill_rating)
                WHERE id = $5 AND tenant_id = $6
            `, [isWin ? 1 : 0, isLoss ? 1 : 0, isDraw ? 1 : 0, newRating, result.id, tenantId]);
        }

        res.json({ message: 'Match results submitted', seasonId });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Helpers
// =========================================

function _getPeriodFilter(period) {
    const now = new Date();
    switch (period) {
        case 'daily':
            now.setHours(0, 0, 0, 0);
            return now.toISOString();
        case 'weekly':
            now.setDate(now.getDate() - now.getDay());
            now.setHours(0, 0, 0, 0);
            return now.toISOString();
        case 'monthly':
            now.setDate(1);
            now.setHours(0, 0, 0, 0);
            return now.toISOString();
        default:
            return '1970-01-01T00:00:00.000Z';
    }
}

module.exports = router;

File: server-rs/src/routes/leaderboards.rs
use axum::{
    extract::{Path, Query, State},
    Json,
};
use serde::Deserialize;
use serde_json::{json, Value};

use crate::error::AppResult;
use crate::middleware::auth::AuthPlayer;
use crate::middleware::tenant::TenantId;
use crate::services::leaderboard;
use crate::AppState;

#[derive(Deserialize)]
pub struct PaginationQuery {
    pub offset: Option<i64>,
    pub limit: Option<i64>,
}

pub async fn get_game_leaderboard(
    State(state): State<AppState>,
    tenant: axum::Extension<TenantId>,
    Path(game_id): Path<String>,
    Query(q): Query<PaginationQuery>,
) -> AppResult<Json<Value>> {
    let tenant_id = &tenant.0 .0;
    let limit = q.limit.unwrap_or(50).min(100) as usize;

    // Try cache first
    let entries = leaderboard::get_top_k(
        &state.cache,
        tenant_id,
        &game_id,
        limit,
        state.config.leaderboard.shard_count,
    )
    .await;

    if !entries.is_empty() {
        let results: Vec<Value> = entries
            .iter()
            .enumerate()
            .map(|(i, (pid, score))| {
                json!({"rank": i + 1, "playerId": pid, "score": *score as i64})
            })
            .collect();
        return Ok(Json(json!({ "entries": results, "source": "cache" })));
    }

    // Fallback to DB
    let rows: Vec<(String, i64, String, i64)> = sqlx::query_as(
        r#"SELECT p.id::text, gp.high_score, p.display_name,
            RANK() OVER (ORDER BY gp.high_score DESC)::bigint as rank
        FROM game_progress gp
        JOIN players p ON p.id = gp.player_id AND p.tenant_id = gp.tenant_id
        WHERE gp.tenant_id = $1 AND gp.game_id = $2
        ORDER BY gp.high_score DESC
        LIMIT $3"#,
    )
    .bind(tenant_id)
    .bind(&game_id)
    .bind(limit as i64)
    .fetch_all(&state.db)
    .await?;

    let results: Vec<Value> = rows
        .iter()
        .map(|(pid, score, name, rank)| {
            json!({"rank": rank, "playerId": pid, "displayName": name, "score": score})
        })
        .collect();

    Ok(Json(json!({ "entries": results, "source": "db" })))
}

pub async fn get_my_rank(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Path(game_id): Path<String>,
) -> AppResult<Json<Value>> {
    let tenant_id = &tenant.0 .0;
    let pid = player.id.to_string();

    // Try cache
    if let Some(rank) = leaderboard::get_approx_rank(
        &state.cache,
        tenant_id,
        &game_id,
        &pid,
        state.config.leaderboard.shard_count,
    )
    .await
    {
        return Ok(Json(json!({"rank": rank, "source": "cache"})));
    }

    // Fallback to DB
    let score: Option<i64> = sqlx::query_scalar(
        "SELECT high_score FROM game_progress WHERE player_id = $1 AND tenant_id = $2 AND game_id = $3",
    )
    .bind(player.id)
    .bind(tenant_id)
    .bind(&game_id)
    .fetch_optional(&state.db)
    .await?;

    match score {
        Some(s) => {
            let rank: i64 = sqlx::query_scalar(
                "SELECT COUNT(*)::bigint + 1 FROM game_progress WHERE tenant_id = $1 AND game_id = $2 AND high_score > $3",
            )
            .bind(tenant_id)
            .bind(&game_id)
            .bind(s)
            .fetch_one(&state.db)
            .await?;
            Ok(Json(json!({"rank": rank, "score": s})))
        }
        None => Ok(Json(json!({"rank": null, "score": 0}))),
    }
}

pub async fn get_around_me(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Path(game_id): Path<String>,
) -> AppResult<Json<Value>> {
    let tenant_id = &tenant.0 .0;

    let my_score: Option<i64> = sqlx::query_scalar(
        "SELECT high_score FROM game_progress WHERE player_id = $1 AND tenant_id = $2 AND game_id = $3",
    )
    .bind(player.id)
    .bind(tenant_id)
    .bind(&game_id)
    .fetch_optional(&state.db)
    .await?;

    let score = match my_score {
        Some(s) => s,
        None => return Ok(Json(json!({"entries": []}))),
    };

    let rows: Vec<(String, i64, String)> = sqlx::query_as(
        r#"(SELECT p.id::text, gp.high_score, p.display_name
            FROM game_progress gp JOIN players p ON p.id = gp.player_id AND p.tenant_id = gp.tenant_id
            WHERE gp.tenant_id = $1 AND gp.game_id = $2 AND gp.high_score > $3
            ORDER BY gp.high_score ASC LIMIT 5)
        UNION ALL
        (SELECT p.id::text, gp.high_score, p.display_name
            FROM game_progress gp JOIN players p ON p.id = gp.player_id AND p.tenant_id = gp.tenant_id
            WHERE gp.tenant_id = $1 AND gp.game_id = $2 AND gp.high_score <= $3
            ORDER BY gp.high_score DESC LIMIT 5)"#,
    )
    .bind(tenant_id)
    .bind(&game_id)
    .bind(score)
    .fetch_all(&state.db)
    .await?;

    let entries: Vec<Value> = rows
        .iter()
        .map(|(pid, s, name)| json!({"playerId": pid, "score": s, "displayName": name}))
        .collect();

    Ok(Json(json!({ "entries": entries })))
}

pub async fn get_global_leaderboard(
    State(state): State<AppState>,
    tenant: axum::Extension<TenantId>,
    Query(q): Query<PaginationQuery>,
) -> AppResult<Json<Value>> {
    let tenant_id = &tenant.0 .0;
    let limit = q.limit.unwrap_or(50).min(100);

    let rows: Vec<(String, i64, String)> = sqlx::query_as(
        "SELECT id::text, total_score, display_name FROM players WHERE tenant_id = $1 ORDER BY total_score DESC LIMIT $2",
    )
    .bind(tenant_id)
    .bind(limit)
    .fetch_all(&state.db)
    .await?;

    let entries: Vec<Value> = rows
        .iter()
        .enumerate()
        .map(|(i, (pid, score, name))| {
            json!({"rank": i + 1, "playerId": pid, "displayName": name, "totalScore": score})
        })
        .collect();

    Ok(Json(json!({ "entries": entries })))
}

pub async fn get_friends_leaderboard(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Path(game_id): Path<String>,
) -> AppResult<Json<Value>> {
    let tenant_id = &tenant.0 .0;

    let rows: Vec<(String, i64, String)> = sqlx::query_as(
        r#"SELECT p.id::text, gp.high_score, p.display_name
        FROM game_progress gp
        JOIN players p ON p.id = gp.player_id AND p.tenant_id = gp.tenant_id
        WHERE gp.tenant_id = $1 AND gp.game_id = $2 AND (
            gp.player_id = $3 OR
            gp.player_id IN (
                SELECT CASE WHEN player_id = $3 THEN friend_id ELSE player_id END
                FROM friendships WHERE tenant_id = $1 AND status = 'accepted'
                    AND (player_id = $3 OR friend_id = $3)
            )
        )
        ORDER BY gp.high_score DESC LIMIT 50"#,
    )
    .bind(tenant_id)
    .bind(&game_id)
    .bind(player.id)
    .fetch_all(&state.db)
    .await?;

    let entries: Vec<Value> = rows
        .iter()
        .enumerate()
        .map(|(i, (pid, score, name))| {
            json!({"rank": i + 1, "playerId": pid, "displayName": name, "score": score})
        })
        .collect();

    Ok(Json(json!({ "entries": entries })))
}

pub async fn get_ranked_leaderboard(
    State(state): State<AppState>,
    tenant: axum::Extension<TenantId>,
    Path(game_id): Path<String>,
    Query(q): Query<PaginationQuery>,
) -> AppResult<Json<Value>> {
    let tenant_id = &tenant.0 .0;
    let limit = q.limit.unwrap_or(50).min(100);

    let rows: Vec<(String, i64, f64, i32, i32)> = sqlx::query_as(
        r#"SELECT le.player_id::text, le.score, le.skill_rating, le.wins, le.matches_played
        FROM leaderboard_entries le
        WHERE le.tenant_id = $1 AND le.game_id = $2 AND le.region = 'global'
        ORDER BY le.skill_rating DESC
        LIMIT $3"#,
    )
    .bind(tenant_id)
    .bind(&game_id)
    .bind(limit)
    .fetch_all(&state.db)
    .await?;

    let entries: Vec<Value> = rows
        .iter()
        .enumerate()
        .map(|(i, (pid, score, rating, wins, matches))| {
            json!({"rank": i + 1, "playerId": pid, "score": score, "skillRating": rating, "wins": wins, "matchesPlayed": matches})
        })
        .collect();

    Ok(Json(json!({ "entries": entries })))
}

pub async fn get_seasons(
    State(state): State<AppState>,
    tenant: axum::Extension<TenantId>,
) -> AppResult<Json<Value>> {
    let rows: Vec<(String, String, chrono::DateTime<chrono::Utc>, chrono::DateTime<chrono::Utc>, bool)> = sqlx::query_as(
        "SELECT id, name, starts_at, ends_at, is_active FROM seasons WHERE tenant_id = $1 ORDER BY starts_at DESC LIMIT 20",
    )
    .bind(&tenant.0 .0)
    .fetch_all(&state.db)
    .await?;

    let seasons: Vec<Value> = rows
        .iter()
        .map(|(id, name, start, end, active)| {
            json!({"id": id, "name": name, "startsAt": start, "endsAt": end, "isActive": active})
        })
        .collect();

    Ok(Json(json!({ "seasons": seasons })))
}

pub async fn get_current_season(
    State(state): State<AppState>,
    tenant: axum::Extension<TenantId>,
) -> AppResult<Json<Value>> {
    let row: Option<(String, String, chrono::DateTime<chrono::Utc>, chrono::DateTime<chrono::Utc>)> = sqlx::query_as(
        "SELECT id, name, starts_at, ends_at FROM seasons WHERE tenant_id = $1 AND is_active = true LIMIT 1",
    )
    .bind(&tenant.0 .0)
    .fetch_optional(&state.db)
    .await?;

    match row {
        Some((id, name, start, end)) => Ok(Json(json!({
            "id": id, "name": name, "startsAt": start, "endsAt": end, "isActive": true
        }))),
        None => Ok(Json(json!({ "season": null }))),
    }
}

pub async fn submit_match(
    State(state): State<AppState>,
    tenant: axum::Extension<TenantId>,
    Json(body): Json<crate::models::multiplayer::SubmitMatchRequest>,
) -> AppResult<Json<Value>> {
    let tenant_id = &tenant.0 .0;

    for pr in &body.players {
        let player_id = uuid::Uuid::parse_str(&pr.player_id)
            .map_err(|_| crate::error::AppError::BadRequest("Invalid player ID".into()))?;

        // Upsert global leaderboard entry
        sqlx::query(
            r#"INSERT INTO leaderboard_entries (tenant_id, player_id, game_id, region, score, wins, losses, draws, matches_played, skill_rating, updated_at)
            VALUES ($1, $2, $3, 'global', $4, $5, $6, 0, 1, 1000, NOW())
            ON CONFLICT (tenant_id, player_id, game_id, region) DO UPDATE SET
                score = leaderboard_entries.score + EXCLUDED.score,
                wins = leaderboard_entries.wins + EXCLUDED.wins,
                losses = leaderboard_entries.losses + EXCLUDED.losses,
                matches_played = leaderboard_entries.matches_played + 1,
                updated_at = NOW()"#,
        )
        .bind(tenant_id)
        .bind(player_id)
        .bind(&body.game_id)
        .bind(pr.score)
        .bind(if pr.is_winner { 1i32 } else { 0 })
        .bind(if pr.is_winner { 0i32 } else { 1 })
        .execute(&state.db)
        .await?;
    }

    Ok(Json(json!({"success": true})))
}
